### PR TITLE
Security Controller

### PR DESCRIPTION
**Pull request by y-abs**
## What does this PR do ?

Add documentation and snippets to existing method

### Other changes
Correct `RoleSearchResult`
Add option `force`

Change signature of:

`updateUserMapping`
`updateRoleMapping`
`updateRole`
`updateProfileMapping`
`mDeleteUsers`
`mDeleteRoles`
`mDeleteProfiles`
`getProfileRights`
`getCredentialFields`
`createUser`
`createRestrictedUser`
`createFirstAdmin`

to make it consistent with sdk JS

### Boyscout

Add `waitForRefresh` option when its missing

### CODE REVIEW

File: doc/2/controllers/security/create-credentials/index.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+The credentials to send. The expected properties depend on the target authentication strategy.
+
+Example for the `local` strategy:
+
+```js
+{
+  username: 'foo',
+  password: 'bar'
+}
```
Comment (jenow) [inline doc/2/controllers/security/create-credentials/index.md]:
```suggestion
```dart
{
  'username': 'foo',
  'password': 'bar'
}
```

File: lib/src/controllers/security.dart [hunk 2]
```diff
@@ -562,58 +567,53 @@ class SecurityController extends KuzzleController {
...
+      body: mapping,
     ));
 
     return response.result as Map<String, dynamic>;
   }
 
   /// Updates a user definition.
-  Future<KuzzleUser> updateUser(Map<String, dynamic> content,
-      {String uid, bool waitForRefresh}) async {
+  Future<KuzzleUser> updateUser(String uid, Map<String, dynamic> body,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
  Future<KuzzleUser> updateUser(String id, Map<String, dynamic> body,
```

File: lib/src/controllers/security.dart [hunk 3]
```diff
@@ -516,13 +521,15 @@ class SecurityController extends KuzzleController {
 
   /// Updates a user credentials for the specified authentication strategy.
   Future<Map<String, dynamic>> updateCredentials(
-      String strategy, String uid, Map<String, dynamic> credentials) async {
+      String strategy, String uid, Map<String, dynamic> credentials,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
      String strategy, String id, Map<String, dynamic> credentials,
```

File: lib/src/controllers/security.dart [hunk 4]
```diff
@@ -451,14 +455,14 @@ class SecurityController extends KuzzleController {
   }
 
   /// Replaces a user with new configuration.
-  Future<KuzzleUser> replaceUser(Map<String, dynamic> content,
-      {String uid, bool waitForRefresh}) async {
+  Future<KuzzleUser> replaceUser(String uid, Map<String, dynamic> body,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
  Future<KuzzleUser> replaceUser(String id, Map<String, dynamic> body,
```

File: lib/src/controllers/security.dart [hunk 5]
```diff
@@ -139,66 +138,70 @@ class SecurityController extends KuzzleController {
...
       uid: uid,
+      waitForRefresh: waitForRefresh,
     ));
 
     return response.result as Map<String, dynamic>;
   }
 
   /// Deletes a user and all their associate credentials.
-  Future<Map<String, dynamic>> deleteUser(String uid) async {
+  Future<Map<String, dynamic>> deleteUser(String uid,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
  Future<Map<String, dynamic>> deleteUser(String id,
```

File: lib/src/controllers/security.dart [hunk 6]
```diff
@@ -139,66 +138,70 @@ class SecurityController extends KuzzleController {
...
+      waitForRefresh: waitForRefresh,
     ));
 
     return response.result as Map<String, dynamic>;
   }
 
   /// Deletes a security role.
   /// An error is returned if the role is still in use.
-  Future<Map<String, dynamic>> deleteRole(String uid) async {
+  Future<Map<String, dynamic>> deleteRole(String uid,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
  Future<Map<String, dynamic>> deleteRole(String id,
```

File: lib/src/controllers/security.dart [hunk 7]
```diff
@@ -139,66 +138,70 @@ class SecurityController extends KuzzleController {
...
+      waitForRefresh: waitForRefresh,
     ));
 
     return response.result as Map<String, dynamic>;
   }
 
   /// Deletes a security profile.
   /// An error is returned if the profile is still in use.
-  Future<Map<String, dynamic>> deleteProfile(String uid) async {
+  Future<Map<String, dynamic>> deleteProfile(String uid,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
  Future<Map<String, dynamic>> deleteProfile(String id,
```

File: lib/src/controllers/security.dart [hunk 8]
```diff
@@ -139,66 +138,70 @@ class SecurityController extends KuzzleController {
...
+      body: body
     ));
 
     return KuzzleUser.fromKuzzleResponse(kuzzle, response);
   }
 
   /// Deletes user credentials for the specified authentication strategy.
   Future<Map<String, dynamic>> deleteCredentials(
-      String strategy, String uid) async {
+      String strategy, String uid, {bool waitForRefresh}) async {
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
      String strategy, String id, {bool waitForRefresh}) async {
```

File: lib/src/controllers/security.dart [hunk 9]
```diff
@@ -139,66 +138,70 @@ class SecurityController extends KuzzleController {
   }
 
   /// Creates a new user
-  Future<KuzzleUser> createUser(
-      Map<String, dynamic> credentials, Map<String, dynamic> content,
-      {String uid, bool waitForRefresh}) async {
+  Future<KuzzleUser> createUser(String uid,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
  Future<KuzzleUser> createUser(String id, 
```

File: lib/src/controllers/security.dart [hunk 10]
```diff
@@ -106,30 +107,28 @@ class SecurityController extends KuzzleController {
   /// This method allows users with limited rights to create other accounts,
   /// but blocks them from creating accounts with unwanted privileges
   /// (e.g. an anonymous user creating his own account).
-  Future<KuzzleUser> createRestrictedUser(Map<String, dynamic> credentials,
-      {Map<String, dynamic> content, String uid, bool waitForRefresh}) async {
+  Future<KuzzleUser> createRestrictedUser(Map<String, dynamic> body, String uid,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
  Future<KuzzleUser> createRestrictedUser(Map<String, dynamic> body, String id,
```

File: lib/src/controllers/security.dart [hunk 11]
```diff
@@ -67,11 +67,12 @@ class SecurityController extends KuzzleController {
   /// identifier already exists, replaces it.
   Future<KuzzleRole> createOrReplaceRole(
       String uid, Map<String, dynamic> controllers,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
      String id, Map<String, dynamic> controllers,
```

File: lib/src/controllers/security.dart [hunk 12]
```diff
@@ -29,17 +32,14 @@ class SecurityController extends KuzzleController {
 
   /// Creates a Kuzzle administrator account, only if none exist.
   Future<KuzzleUser> createFirstAdmin(
-      String uid, Map<String, dynamic> credentials,
-      {Map<String, dynamic> content, bool reset}) async {
+      String uid, Map<String, dynamic> body,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
      String id, Map<String, dynamic> body,
```

File: lib/src/controllers/security.dart [hunk 13]
```diff
@@ -15,7 +15,10 @@ class SecurityController extends KuzzleController {
 
   /// Creates authentication credentials for a user.
   Future<Map<String, dynamic>> createCredentials(
-      String strategy, String uid, Map<String, dynamic> credentials) async {
+      String strategy,
+      String uid,
```
Comment (jenow) [inline lib/src/controllers/security.dart]:
```suggestion
      String id,
```

File: doc/2/controllers/security/create-role/index.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+---
+
+# createRole
+
+Creates a new role.
+
+<br />
+
+```dart
+Future<KuzzleRole> createRole(String uid, Map<String, dynamic> controllers,
```
Comment (jenow) [inline doc/2/controllers/security/create-role/index.md]:
```suggestion
Future<KuzzleRole> createRole(String id, Map<String, dynamic> controllers,
```

File: doc/2/controllers/security/delete-credentials/index.md [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+# deleteCredentials
+
+Deletes user credentials for the specified authentication strategy.
+
+<br />
+
+```dart
+Future<Map<String, dynamic>> deleteCredentials(
+      String strategy, String uid, {bool waitForRefresh})
```
Comment (jenow) [inline doc/2/controllers/security/delete-credentials/index.md]:
```suggestion
      String strategy, String id, {bool waitForRefresh})
```

File: doc/2/controllers/security/replace-user/index.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+---
+
+# replaceUser
+
+Replaces a user with new configuration.
+
+<br />
+
+```dart
+Future<KuzzleUser> replaceUser(String uid,
```
Comment (jenow) [inline doc/2/controllers/security/replace-user/index.md]:
```suggestion
Future<KuzzleUser> replaceUser(String id, 
```

File: doc/2/controllers/security/update-credentials/index.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+# updateCredentials
+
+Updates a user credentials for the specified authentication strategy.
+
+<br />
+
+```dart
+ Future<Map<String, dynamic>> updateCredentials(
+      String strategy, String uid, Map<String, dynamic> credentials,
```
Comment (jenow) [inline doc/2/controllers/security/update-credentials/index.md]:
```suggestion
      String strategy, String id, Map<String, dynamic> credentials,
```

File: doc/2/controllers/security/update-profile/index.md [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+---
+
+# updateProfile
+
+Updates a security profile definition.
+
+<br />
+
+```dart
+ Future<KuzzleProfile> updateProfile(String uid, List<dynamic> policies,
```
Comment (jenow) [inline doc/2/controllers/security/update-profile/index.md]:
```suggestion
 Future<KuzzleProfile> updateProfile(String id, List<dynamic> policies,
```

File: doc/2/controllers/security/update-role/index.md [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+# updateRole
+
+Updates a security role definition.
+
+**Note**: partial updates are not supported for roles, this API route will replace the entire role content with the provided one.
+
+<br />
+
+```dart
+Future<KuzzleRole> updateRole(String uid, Map<String, dynamic> controllers,
```
Comment (jenow) [inline doc/2/controllers/security/update-role/index.md]:
```suggestion
Future<KuzzleRole> updateRole(String id, Map<String, dynamic> controllers,
```

File: doc/2/controllers/security/update-user/index.md [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+---
+
+# updateUser
+
+Updates a user definition.
+
+<br />
+
+```dart
+Future<KuzzleUser> updateUser(String uid, Map<String, dynamic> body,
```
Comment (jenow) [inline doc/2/controllers/security/update-user/index.md]:
```suggestion
Future<KuzzleUser> updateUser(String id, Map<String, dynamic> body,
```

File: doc/2/controllers/security/create-user/index.md [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+---
+
+# createUser
+
+Creates a new user.
+
+<br />
+
+```dart
+Future<KuzzleUser> createUser(String uid,
```
Comment (jenow) [inline doc/2/controllers/security/create-user/index.md]:
```suggestion
Future<KuzzleUser> createUser(String id, 
```

File: doc/2/controllers/security/create-user/index.md [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+Future<KuzzleUser> createUser(String uid, 
+      Map<String, dynamic> body,
+      {bool waitForRefresh})
+```
+
+<br />
+
+| Property | Type | Description |
+|--- |--- |--- |
+| `kuid` | <pre>String</pre> | User [kuid](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```
Comment (jenow) [inline doc/2/controllers/security/create-user/index.md]:
```suggestion
| `id` | <pre>String</pre> | User [id](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```

File: doc/2/controllers/security/delete-credentials/index.md [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+```
+
+<br />
+
+| Property | Type | Description |
+| --- | --- | --- |
+| `strategy` | <pre>String</pre> | Strategy to use |
+| `kuid` | <pre>String</pre> | User [kuid](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```
Comment (jenow) [inline doc/2/controllers/security/delete-credentials/index.md]:
```suggestion
| `id` | <pre>String</pre> | User [id](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```

File: doc/2/controllers/security/delete-profile/snippets/delete-profile.dart [hunk 24]
```diff
@@ -0,0 +1,11 @@
+final result = await kuzzle.security.deleteProfile('myProfile');
+print('Profile successfully deleted');
```
Comment (jenow) [inline doc/2/controllers/security/delete-profile/snippets/delete-profile.dart]: Please do the print in the template by using the template `default` or `print-result` for example.

File: doc/2/controllers/security/delete-role/snippets/delete-role.dart [hunk 25]
```diff
@@ -0,0 +1,12 @@
...
+{ found: true,
+  _index: '%kuzzle',
+  _type; 'roles',
+  _id: 'myRole',,
+  _version: 2,
+  result: 'deleted',
+  _shards: {  total: 2, successful: 1, failed: 0 } }
+  */
+
+print('Role successfully deleted');
```
Comment (jenow) [inline doc/2/controllers/security/delete-role/snippets/delete-role.dart]: Please do the print in the template by using the template `default` or `print-result` for example.

File: doc/2/controllers/security/m-delete-users/index.md [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+```dart
+ Future<List<dynamic>> mDeleteUsers(List<String> ids,
+      {bool waitForRefresh})
+```
+
+<br />
+
+| Property | Type | Description |
+|--- |--- |--- |
+| `kuids` | <pre>List<String></pre> | Array of user [kuid](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```
Comment (jenow) [inline doc/2/controllers/security/m-delete-users/index.md]:
```suggestion
| `ids` | <pre>List<String></pre> | Array of user [id](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```

File: doc/2/controllers/security/replace-user/index.md [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+```dart
+Future<KuzzleUser> replaceUser(String uid, 
+  Map<String, dynamic> body, {bool waitForRefresh})
+```
+
+<br />
+
+| Property | Type | Description |
+|--- |--- |--- |
+| `kuid` | <pre>String</pre> | User [kuid](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```
Comment (jenow) [inline doc/2/controllers/security/replace-user/index.md]:
```suggestion
| `id` | <pre>String</pre> | User [id](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```

File: doc/2/controllers/security/update-credentials/index.md [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+      {bool waitForRefresh})
+```
+
+<br />
+
+| Property | Type | Description |
+| --- | --- | --- |
+| `strategy` | <pre>String</pre> | Strategy to use |
+| `kuid` | <pre>String</pre> | User [kuid](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```
Comment (jenow) [inline doc/2/controllers/security/update-credentials/index.md]:
```suggestion
| `id` | <pre>String</pre> | User [id](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```

File: doc/2/controllers/security/update-user/index.md [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+      {bool waitForRefresh})
+```
+
+<br />
+
+| Property | Type | Description |
+|--- |--- |--- |
+| `kuid` | <pre>String</pre> | User [kuid](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```
Comment (jenow) [inline doc/2/controllers/security/update-user/index.md]:
```suggestion
| `id` | <pre>String</pre> | User [id](/core/2/guides/essentials/user-authentication#kuzzle-user-identifier-kuid) |
```

**Review by jenow — CHANGES_REQUESTED** [review]

**Review by jenow — APPROVED** [review]

**Review by alexandrebouthinon — APPROVED** [review]

**Review by alexandrebouthinon — APPROVED** [review]

**Pull request merged**